### PR TITLE
Replace custom IOHW -> OIHW reorder with build-in oneDNN reorder

### LINKED ISSUE #18842: [MKL-DNN] Replace manually implemented reorder iohw->oihw with mkl-dnn one
At the time implementing conv_transpose mkl-dnn mkl-dnn did not recognize iohw format hence
custom reorder was implemeted on paddle side. Later on mkl-dnn added iohw format and now we could remove custom reorder.

This is cleanup task. For inference this custom reorder is called only at the beginning.

### PR DESCRIPTION
**Pull request by Silv3S**
### PR types
Performance optimization

### PR changes
OPs

### Describe
PR resolves #18842

### CODE REVIEW

File: paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc [hunk 1]
```diff
@@ -229,32 +229,21 @@ class ConvTransposeMKLDNNHandlerT
...
-          int in_offset = j * h * w + i * o * h * w;
-          int out_offset = j * c * h * w + i * h * w;
-          std::memcpy(&(reordered_filter_data.get())[out_offset],
-                      &filter_data[in_offset], h * w * sizeof(K));
-        }
-      }
-
-      return reordered_filter_data;
-    };
+    platform::MKLDNNEngine engine = dev_ctx.GetEngine();
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc]:
```suggestion
    const platform::MKLDNNEngine& engine = dev_ctx.GetEngine();
```

File: paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc [hunk 2]
```diff
@@ -229,32 +229,21 @@ class ConvTransposeMKLDNNHandlerT
...
+    platform::MKLDNNStream engine_stream(engine);
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc]:
```suggestion
    auto& astream = platform::MKLDNNDeviceContext::tls().get_stream();
```

File: paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc [hunk 3]
```diff
@@ -229,32 +229,21 @@ class ConvTransposeMKLDNNHandlerT
...
+        weights_tz, platform::MKLDNNGetDataType<K>(), MKLDNNMemoryFormat::iohw);
+    auto reordered_data_md = platform::MKLDNNMemDesc(
+        weights_tz, platform::MKLDNNGetDataType<K>(), MKLDNNMemoryFormat::oihw);
+    auto source_data_mem = platform::MKLDNNMemory(
+        source_data_md, engine, platform::to_void_cast<K>(filter_data));
+    auto reordered_data_mem = platform::MKLDNNMemory(reordered_data_md, engine);
+    platform::Reorder(source_data_mem, reordered_data_mem, engine);
+    engine_stream.wait();
 
     return this->template AcquireMemoryWithReorder<K>(
```
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc:232]: Why "AcquireMemoryWithReorder" is called just after doing a reorder?

**Review by jakpiase — COMMENTED** [review]
Nicely done, but don't you want to try out AcquireApi?

File: paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc [hunk 4]
```diff
@@ -229,32 +229,21 @@ class ConvTransposeMKLDNNHandlerT
...
-          int in_offset = j * h * w + i * o * h * w;
-          int out_offset = j * c * h * w + i * h * w;
-          std::memcpy(&(reordered_filter_data.get())[out_offset],
-                      &filter_data[in_offset], h * w * sizeof(K));
-        }
-      }
-
-      return reordered_filter_data;
-    };
+    const platform::MKLDNNEngine& engine = dev_ctx.GetEngine();
```
Comment (jczaja) [inline paddle/fluid/operators/mkldnn/conv_transpose_mkldnn_op.cc]: I actually think  this block of code is not needed. What should be modified is fragment "filter->format()" to "dnnl::format_tag::iohw" (guess) and then AcquireMemoryWithReorder will take care of rordering from IOHW -> blocked_optimized format. In your current changes you have two reorders: IOHW -> OIHW and then OIHW to blocked_optimized_format.  One issue here is that filter-<format() is returning (probably) OIHW while in fact data is IOHW .

**Review by jczaja — APPROVED** [review]
LGTM

**Review by jakpiase — APPROVED** [review]
LGTM, good job!

**Comment by Silv3S** [comment]
@jczaja please merge

**Pull request merged**